hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java

```diff
@@ -612,6 +612,9 @@ public boolean next(List<Cell> outResult, ScannerContext scannerContext) throws
           // here, we still need to scan all the qualifiers before returning...
           scannerContext.returnImmediately();
         }
+
+        scannerContext.incrementBlockProgress(heap.getCurrentBlockSizeOnce());
```

Comment:
So the problem here is we do not know if this is a new block?

Reply:
Correct. KeyValueHeap makes this tricky, this is why the API design is strange.  In the context of a single `StoreScanner.next` call, we can jump back and forth between different blocks in different StoreFiles. This approach here ensures that we always only count each block once.

I considered a few different approaches, and I felt this one was the simplest:

1. Similar to RSRpcServices.addSize, try tracking blocks here by checking `heap.getCurrentBlock() != lastBlock`. As mentioned above, this is error prone (as the comments in addSize mention as well).
2. Try adding an `HFileReaderImpl.setBlockChangedCallback(newBlock -> scannerContext.incrementBlockSize(newBlock))`.  This would work, but opens up issue with how StoreScanner can re-open StoreFileScanners due to flushes or read type switches, etc. We'd need to update this callback in RegionScannerImpl.nextInternal, which would need to delegate out to all StoreScanners -> KeyValueHeaps -> StoreFileScanners.
3. Add a `HFileReaderImpl.getBlockChanged()` and `HFileReader.getCurrentBlockSize()`. Then we could change this line to:

```java
if (heap.getBlockChanged()) {
  scannerContext.incrementBlockProgress(heap.getCurrentBlockSize());
}
```

The implementation of HFileReaderImpl.getBlockChanged would be:

```java
public boolean getBlockChanged() {
  if (blockChanged) {
    blockChanged = false;
    return true;
  }
  return false;
}
```

We'd reset blockChanged to true whenever we update curBlock in HFileReaderImpl.

Option 3 would probably be the closest to what I have and most accurate/simplest, and it might be more intuitive API as well. The only downside would be 2 new interface methods instead of 1.

Would you like me to do Option 3?

Reply:
@Apache9 here's a commit which implements the `getBlockChanged()` approach: https://github.com/HubSpot/hbase/commit/c619b4f3f7cd7f9e9677b265e10f0f10edefcf23

Let me know if you'd like me to push that here instead. It's better in some ways, but still has one unfortunate caveat described in the javadoc of KeyValueScanner. See the "Note:" portion here:

```java
/**
   * Returns true if the current block has changed since last invocation. Subsequent calls to this
   * method will return false, until the block changes again. <br>
   * Note: This value is propagated up from the HFile.Reader. The value of this result is only valid
   * until the next call to {@link #next()} or any other seek-related method. Calling those methods
   * may change which underlying HFile is currently being read, which will reset the block changed
   * state.
   */
```

This is ok for our current use-case, since we always invoke the 2 methods together. But in some ways it would be better to combine these 2 methods into one (like in `getCurrentBlockSizeOnce()`) so that you can't stumble across that note.